### PR TITLE
fix #51135 make CFS quota period configurable

### LINKED ISSUE #51135: Avoid setting CPU limits for Guaranteed pods
The effect of CPU throttling is non obvious to users and throws them off when they try out kubernetes. It also complicates CPU capacity planning for pods. Pods that are carefully placed in Guaranteed QoS class are typically of higher priority than other pods, and these pods are expected to get the best performance. On the other hand, Burstable pods may have to be limited in certain environments to avoid tail latencies for Guaranteed pods.
I propose we avoid setting CPU limits on Guaranteed pods by default and continue to impose CPU limits on Burstable pods. We can provide a config option to ease the rollout in existing clusters. Users may need time to handle this transition, and some of them may choose to use the existing policies. 

Closely related is static CPU pinning policies in kubelet which will naturally limit Guaranteed pods to specific cores if they request integral number of cores. This feature may be combined with turning off CPU limiting for Guaranteed pods. That said, having an ability to configure each of these feature independently will still be helpful.

@derekwaynecarr we have discussed this in the past, but AFAIK, an issue hasn't been filed thus far. @dchen1107 thoughts? 

If there is a general agreement that this is the intended direction longer term, I can post a patch to make CPU limit policies configurable.

@kubernetes/sig-node-feature-requests

### PR DESCRIPTION
**Pull request by szuecs**
**What this PR does / why we need it**:

This PR makes it possible for users to change CFS quota period from the default 100ms to some other value between 1µs and 1s.
#51135 shows that multiple production users have serious issues running reasonable workloads in kubernetes. The latency added by the 100ms CFS quota period is adding way too much time.

**Which issue(s) this PR fixes**:
Fixes #51135 

**Special notes for your reviewer**:
- 5ms is used by user experience https://github.com/kubernetes/kubernetes/issues/51135#issuecomment-384908627
- Latency added caused by CFS 100ms is shown at https://github.com/kubernetes/kubernetes/issues/51135#issuecomment-373454012
- explanation why we should not disable limits https://github.com/kubernetes/kubernetes/issues/51135#issuecomment-385346661
- agreement found at kubecon EU 2018: https://github.com/kubernetes/kubernetes/issues/51135#issuecomment-386623964

**Release note**:
```release-note
Adds a kubelet parameter and config option to change CFS quota period from the default 100ms to some other value between 1µs and 1s. This was done to improve response latencies for workloads running in clusters with guaranteed and burstable QoS classes.  
```

### CODE REVIEW

**Comment by liggitt** [comment]
@kubernetes/sig-node-pr-reviews  @derekwaynecarr

**Comment by dims** [comment]
@derekwaynecarr The `"cpu-cfs-quota"` command line flag in kubelet right? So we would have a `"cpu-cfs-quota-period"` to go along with it?

**Comment by derekwaynecarr** [comment]
Yes

**Comment by dims** [comment]
@szuecs looking forward to an update in this PR with the suggestion above

**Comment by derekwaynecarr** [comment]
see my comment here for why a flag is preferred.  Not all nodes are tuned for latency.

https://github.com/kubernetes/kubernetes/issues/51135#issuecomment-386767390

**Comment by szuecs** [comment]
Thanks for all your comments.

**Comment by szuecs** [comment]
@derekwaynecarr I added the cli flag and config option. I hope that I added it everywhere, where it is needed.

I tested, that:
- tests run and report ok % go test ./pkg/kubelet/cm -v 
- I can locally build kubelet: % make WHAT=cmd/kubelet

Let me know if I have to change anything.

**Comment by vishh** [comment]
/hold
I'd like for us to reach a consensus on https://github.com/kubernetes/kubernetes/issues/51135#issuecomment-387139466 prior to merging this patch.

**Comment by dims** [comment]
@derekwaynecarr @vishh Do we count this as a new feature? if so, Does this feature need an alpha feature gate?

**Comment by dims** [comment]
@derekwaynecarr @vishh Also, do we leave the current default as-is? (when nothing is specified in the command line)

**Comment by dims** [comment]
@szuecs 

- if run "hack/update-bazel.sh" that should take care of the verify job failure. (and bazel-test job failure too)
- the fix for e2e failure should have already gone in, but you may have to rebase to master

So, please squash the commits (for easier review), rebase to master and we should get all green (cross my fingers)

thanks,
Dims

**Comment by szuecs** [comment]
@dims The update-bazel.sh did no change anything.
 
Without run-in-gopath.sh I get panics all the time, as far as I understood the developer docs I should run it with run-in-gopath.sh. 
I fetched master and merged into the branch, such that is up to date.
I also pulled the godeps with hack/run-in-gopath.sh hack/godep-restore.sh, but it does not change the update-bazel run:

```
[sszuecs@sandor-lab:~/go/src/k8s.io/kubernetes]% hack/run-in-gopath.sh hack/update-bazel.sh
[sszuecs@sandor-lab:~/go/src/k8s.io/kubernetes]% git st
On branch fix/51135-set-saneer-default-cpu.cfs_period
nothing to commit, working directory clean
```

Any hints?

**Comment by dims** [comment]
/assign @derekwaynecarr 
/assign @vishh 
/assign @liggitt 

Questions:
- Do we need a alpha feature gate?
- Anything specific we need to do for an upgrade scenario?

**Review by dims — APPROVED** [review]
Looks good to me so far. thanks for the rebase + squash.

**Review by timothysc — APPROVED** [review]
/approve

**Comment by derekwaynecarr** [comment]
I am in favor of this change, but would like to discuss this topic in tomorrow's sig-node meeting in order to reach consensus on the path forward.

/cc @dchen1107

**Comment by derekwaynecarr** [comment]
@dims - this should not demand a feature gate.

File: cmd/kubelet/app/options/options.go
Comment (derekwaynecarr) [inline cmd/kubelet/app/options/options.go]: please add that this is the kernel default.

Comment (liggitt) [reply cmd/kubelet/app/options/options.go]: Default is printed automatically, no need to double include in help text

Comment (szuecs) [reply cmd/kubelet/app/options/options.go]: Changed as commented

**Comment by szuecs** [comment]
@derekwaynecarr I think you have to add your approval, too as far as I understand. Let me know if there is something missing or you just need time to review. Thanks

**Comment by derekwaynecarr** [comment]
@szuecs - i will review the code in depth later today or tomorrow.

**Comment by derekwaynecarr** [comment]
Hmm, the plot thickens on this PR.

It appears this PR works with the `cgroupfs` cgroup driver and not the `systemd` cgroup driver.

The issue is that when using systemd, systemd only exposes in its dbus API `CPUQuotaPerSecUSec`.

see in runc:
https://github.com/opencontainers/runc/blob/master/libcontainer/cgroups/systemd/apply_systemd.go#L314

and in systemd the period is fixed via constant:
https://github.com/systemd/systemd/blob/81183d9b9934bc508820d7bedb82695c5f74ed27/src/core/cgroup.c#L517

i will reach out to systemd community to see if i am missing something, but we should think about how to proceed with this pr.  if we ever wanted to support a per pod period tuning option, i would want to have that knob settable via systemd, or know its safe to set it behind systemd before merging this since the systemd cgroup driver is GA.

in the interim, we may want to mark this flag as experimental, and disable the ability to update the value when using systemd cgroup driver.

**Comment by derekwaynecarr** [comment]
I have opened the following issue with systemd: https://github.com/systemd/systemd/issues/9081

**Comment by hjacobs** [comment]
@derekwaynecarr @BenPope thanks! @szuecs could you take a look at https://github.com/szuecs/kubernetes/pull/1?

**Comment by szuecs** [comment]
@BenPope I merged your PR into this PR and squashed the commit. Thanks!

**Comment by dims** [comment]
@szuecs looks like there are some legit failures in the ci jobs

**Comment by szuecs** [comment]
@dims thanks I will check it later today or at least this week. It's not always clear to me when I just have to retest and when it's a real issue.

**Comment by derekwaynecarr** [comment]
it looks like this is a real issue, we can plan for this change in 1.12

**Comment by szuecs** [comment]
@derekwaynecarr fixed

**Comment by dims** [comment]
/hold cancel

/milestone v1.12 

( per last comment from @derekwaynecarr )

**Comment by szuecs** [comment]
kind/feature

**Comment by dims** [comment]
/assign @derekwaynecarr

let's please get this in early in the cycle now that the master is open

File: cmd/kubelet/app/options/options.go
Comment (feiskyer) [inline cmd/kubelet/app/options/options.go]: `defaults to Linux Kernel default (100ms)` for clear?

**Review by feiskyer — COMMENTED** [review]
LGTM

File: cmd/kubelet/app/options/options.go
Comment (szuecs) [reply cmd/kubelet/app/options/options.go]: See the former comment from ligitt. Cobra will render the default automatically.

Comment (dixudx) [reply cmd/kubelet/app/options/options.go]: We should stop adding new flags to kubelet. Add directly to `KubeletConfiguration` instead?

**Review by dixudx — CHANGES_REQUESTED** [review]

**Comment by szuecs** [comment]
@derekwaynecarr @vishh anything that stops this PR from getting a lgtm?
I feel a bit sick of rebasing and fixing merge conflicts and tests, because of refactorings in other places.

**Comment by derekwaynecarr** [comment]
Can we put this behind a feature gate flag?  The systemd community has prepared a patch to add support for variable quota , but since this option is not yet supported across all cgroup drivers, I would prefer to put it behind a feature gate flag in the interim.

**Comment by szuecs** [comment]
@derekwaynecarr can you roughly explain where this is done?
I guess it's a  value in a kubelet flag, right?

**Comment by dims** [comment]
@szuecs you can follow the same pattern as [`SupportPodPidsLimit`](https://cs.k8s.io/?q=SupportPodPidsLimit&i=nope&files=&repos=kubernetes)

**Comment by szuecs** [comment]
I added feature gate CustomCPUCFSQuotaPeriod to enable the feature. 
I will check tomorrow, if the tests are passing.

File: pkg/kubelet/kuberuntime/helpers_linux.go
Comment (dims) [inline pkg/kubelet/kuberuntime/helpers_linux.go]: @szuecs Don't we have to use old behavior here when the alpha feature flag is not set?

```
	if !utilfeature.DefaultFeatureGate.Enabled(kubefeatures.CPUCFSQuotaPeriod) {
		period = quotaPeriod
	}
```

File: pkg/kubelet/kuberuntime/helpers_linux.go
Comment (dims) [inline pkg/kubelet/kuberuntime/helpers_linux.go]: We need this below

File: pkg/kubelet/kuberuntime/helpers_linux.go
Comment (szuecs) [reply pkg/kubelet/kuberuntime/helpers_linux.go]: I added additional unit tests to make sure we have the right values if feature gate is enabled or disabled

**Comment by dims** [comment]
/test pull-kubernetes-e2e-gke
/test pull-kubernetes-cross
/test pull-kubernetes-e2e-kubeadm-gce
/test pull-kubernetes-local-e2e

Run the skipped tests just to be sure!

**Comment by szuecs** [comment]
@derekwaynecarr PR is ready for review, feature gates set up, more unit tests added, all e2e tests green.

**Review by timothysc — APPROVED** [review]
/approve - from my side

File: cmd/kubelet/app/options/options.go
Comment (derekwaynecarr) [reply cmd/kubelet/app/options/options.go]: i would prefer to not have the flag and just use via `KubeletConfiguration`; we should note that dynamically changing the field value will not impact currently running pods (only new pods)

File: pkg/kubelet/apis/kubeletconfig/v1beta1/types.go
Comment (derekwaynecarr) [inline pkg/kubelet/apis/kubeletconfig/v1beta1/types.go]: can you add a note about the impact of changing this field when using Dynamic Kubelet Config?

File: pkg/kubelet/apis/kubeletconfig/v1beta1/defaults.go
Comment (derekwaynecarr) [inline pkg/kubelet/apis/kubeletconfig/v1beta1/defaults.go]: reference the feature gate name from features package directly.

File: pkg/kubelet/apis/kubeletconfig/validation/validation.go
Comment (derekwaynecarr) [inline pkg/kubelet/apis/kubeletconfig/validation/validation.go]: same here reference `features.` feature name directly.

File: pkg/kubelet/cm/helpers_linux_test.go
Comment (derekwaynecarr) [inline pkg/kubelet/cm/helpers_linux_test.go]: doesn't this stop us from testing the function when the feature gate is off?

shouldn't you have a seperate Test function that runs the test with it off, and another with it on?

File: pkg/kubelet/kuberuntime/helpers_linux.go
Comment (derekwaynecarr) [inline pkg/kubelet/kuberuntime/helpers_linux.go]: remove this debugging

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
i have a number of code nits, but how do you feel about writing a node e2e test as well that validates the function?  the test can run if not using systemd cgroup driver, and work similar to pods_container_manager_test ?

File: pkg/kubelet/cm/helpers_linux_test.go
Comment (szuecs) [reply pkg/kubelet/cm/helpers_linux_test.go]: I duplicated the test and have one enabled and one disabled.
One problem for the test is that the ResourceConfigForPod() gets the period as paramter so the test will have expected the tunedQuotaPeriod because of that. The period parameter will not be passed into it, but this is outside of the cm package. 
I am not sure if this is nice to read in the test, maybe you have a better suggestion? 
I could add the feature gate in cm, such that the parameter will be ignored, which I think makes even less sense.

**Comment by szuecs** [comment]
@derekwaynecarr I changed as commented, but see https://github.com/kubernetes/kubernetes/pull/63437#discussion_r212381282, which could be odd. 
I did not add the e2e test, yet.

File: pkg/kubelet/apis/kubeletconfig/v1beta1/types.go
Comment (szuecs) [reply pkg/kubelet/apis/kubeletconfig/v1beta1/types.go]: done

**Comment by derekwaynecarr** [comment]
thank you for the the updated test case, i would prefer an e2e test, but i wont hold this pr on that.

if we ever do graduate the feature to beta, it should definitely have tests.

/approve
/lgtm

**Review by derekwaynecarr — APPROVED** [review]

**Comment by dims** [comment]
reapplying lgtm from @derekwaynecarr (lost in rebase)

/lgtm

**Comment by hjacobs** [comment]
@derekwaynecarr @dims @timothysc @szuecs thanks!

**Comment by dashpole** [comment]
This has broken the release blocking alphafeatures test suite: https://k8s-testgrid.appspot.com/sig-release-1.12-blocking#gce-cos-1.12-alphafeatures

The kubelet is crashlooping:
server.go:207] invalid configuration: CPUCFSQuotaPeriod (--cpu-cfs-quota-period) {0s} must be between 1usec and 1sec, inclusive

Tracking bug for release-blocking failures: https://github.com/kubernetes/kubernetes/issues/68313

File: pkg/kubelet/apis/config/v1beta1/defaults.go
Comment (dashpole) [inline pkg/kubelet/apis/config/v1beta1/defaults.go]: I don't know if this actually works...  We generally do not feature gate the defaulting of flags.  We generally only place the feature gate around code we don't want to run, rather than around configuration.

**Comment by guineveresaenger** [comment]
@dashpole @derekwaynecarr @dims please give this your immediate attention re: #68313.

File: pkg/kubelet/apis/config/v1beta1/defaults.go
Comment (derekwaynecarr) [reply pkg/kubelet/apis/config/v1beta1/defaults.go]: This does look like a problem.  The flag should always default, it just shouldn’t have been used.  Apologies for missing in review

**Comment by dashpole** [comment]
Fix is out: https://github.com/kubernetes/kubernetes/pull/68386

**Comment by warmchang** [comment]
ping @Pingan2017

**Comment by bartoszhernas** [comment]
Our php container started seeing problems connected with CPU limits.

We started profiling and saw that `initSoapClient` method which usually takes 10ms, started taking up to 120s (no mistake, 2 minutes...). We do not use any container limits, we only use requests. After requesting for much more CPU than possibly needed (40%) the issue disappeared.

We are hosted on GKE, is there an option to disable cpu limits or reduce CFS period there? Does anyone has idea how to fix this problem on GKE?

File: pkg/kubelet/apis/config/types.go
Comment (paskal) [inline pkg/kubelet/apis/config/types.go:223]: I'm super confused about this one and can't figure this one out by myself: `cpu.cfs_period_us` default limit called "100ms" but [actually](https://www.kernel.org/doc/html/latest/scheduler/sched-bwc.html#management) is microseconds:
> `cpu.cfs_period_us`: the length of a period (in microseconds)

In the change I'm commenting author clearly acted under assumptions that milliseconds are used in k8s and now it's all over the test cases and comments. But I wonder where that 1000x disperancy between k8s value and Linux value comes from? And does it exist, or k8s default is actually a microsecond, same as linux? Unless we multiple [that](https://github.com/kubernetes/kubernetes/blob/37761a3/pkg/kubelet/cm/helpers_linux.go#L47-L48) default value of `QuotaPeriod` by 1000 somewhere, converted to `time.Duration` it's 100 microseconds and not milliseconds.

cc @vishh @sjenning and especially @hjacobs: I hope you all have some context about it and can tell me why I'm wrong and k8s code comments are right.

Comment (szuecs) [reply pkg/kubelet/apis/config/types.go:223]: yes cfs_period_us is in µs, that's why it has `_us` suffix
Maybe https://go.dev/play/p/65fG1OmLFYN helps to understand.
As far as I remember, I think the time.Duration is only used if you want to set it via kubelet flag/config.

Comment (paskal) [reply pkg/kubelet/apis/config/types.go:223]: The comment in that MR says "ms"." In contrast, it better be "microseconds", I assume. I can send MR to change it as well as explicit millisecond mentions in the current code, whereas it should be microseconds. I'll ping you in it.

In this MR, you wrote milliseconds duration in multiple places in defaults and tests `CPUCFSQuotaPeriod: metav1.Duration{Duration: 100 * time.Millisecond},`, I wonder if that should be changed to microseconds to match the actual default value of 100 microseconds?

Also, the discussion which led to this MR has multiple people talking about changing the value from 100 milliseconds to 5/25 milliseconds for testing. Still, it doesn't make sense unless everyone were altering the source code in microseconds, thinking it's milliseconds. @hjacobs even did [a talk](https://youtu.be/eBChCFD9hfs?t=870) about it talking about the default value as if it was in milliseconds. Your code introducing milliseconds and not microseconds (so 1000x) contributed to my confusion, and I wonder if it's intentional or if you are mistaken, and it should be changed microseconds to change the real default.

The only way current code would make sense to me is if the value you feed to k8s (100 milliseconds) is divided by 1000 at some point in time and then matches the real Linux default of 100 microseconds, but I haven't found such a place yet, and I would doubt that's the case.

Comment (paskal) [reply pkg/kubelet/apis/config/types.go:223]: @szuecs, the simplest form of the question I can form is the following.

Is the 100ms value of `CPUCFSQuota` result in the same k8s behaviour as the default Linux value of 100 microseconds, or is it 1000 times higher as its milliseconds versus microseconds in Linux?

File: pkg/kubelet/kuberuntime/helpers_linux.go
Comment (paskal) [inline pkg/kubelet/kuberuntime/helpers_linux.go:63]: @szuecs that's precise place which worries me: previously we used `quotaPeriod` as-is, which, translated to `time.Duration`, equals to 100 microseconds. And now we set it to `cpuCFSQuotaPeriod`, which according to code and comments is 100 millisecond default.

File: pkg/kubelet/kuberuntime/helpers_linux.go
Comment (szuecs) [inline pkg/kubelet/kuberuntime/helpers_linux.go:34]: quotaPeriod calculation was inlined:
100 * minQuotaPeriod = 100 * 1000 = 100000

File: pkg/kubelet/kuberuntime/helpers_linux.go
Comment (szuecs) [reply pkg/kubelet/kuberuntime/helpers_linux.go:63]: 1. I don't see how this code can be worrying, because it shows the same assignment as before. If not the feature gate is set, set it to quotaPeriod as before. see https://github.com/kubernetes/kubernetes/pull/63437/files/588d2808b77d11f235b6eba5c21bcaa89a2f7804#r932625167
2. if you enable the featuregate you can set to the value you like

Comment (paskal) [reply pkg/kubelet/kuberuntime/helpers_linux.go:63]: "ms" mentioned across the code alongside the tests with milliseconds and not microseconds confused me. Now I see it's still microseconds, and comments should be altered, and maybe tests as well. I'll prepare the PR, thanks for the help!

Comment (szuecs) [reply pkg/kubelet/kuberuntime/helpers_linux.go:63]: Thanks @paskal , makes sense to update, if it's wrong :)

Comment (paskal) [reply pkg/kubelet/kuberuntime/helpers_linux.go:63]: I've created #111520 about it.

Comment (paskal) [reply pkg/kubelet/kuberuntime/helpers_linux.go:63]: I was wrong, and the original code was correct. 100ms is indeed a default value in Linux, and I got confused by conversions in two places in code that had no comments about them. PR #112123 adds comments to the code added here.

File: pkg/kubelet/apis/config/types.go
Comment (paskal) [reply pkg/kubelet/apis/config/types.go:223]: I was wrong, and the original code was correct. 100ms is indeed a default value in Linux, and I got confused by conversions in two places in code that had no comments about them. PR #112123 adds comments to the code added here.

File: pkg/kubelet/kuberuntime/helpers_linux.go
Comment (szuecs) [reply pkg/kubelet/kuberuntime/helpers_linux.go:63]: I think I read the kernel documentation at least 5 times when I did investigation and the code here.

**Pull request merged**
